### ISSUE TITLE
Node.js uses things from <uv.h> that aren't documented. Bug in Node.js or just lacking documentation?

### ISSUE THREAD

**Issue by EdSchouten** [libuv/libuv#1607]
Hi there,

For [CloudABI](https://github.com/NuxiNL/cloudlibc) I am adding a libuv compatible event loop directly to the C library. The official libuv can't be used here for a couple of reasons:

- A fair number of interfaces provided by libuv (signal handling, global filesystem namespace, etc.) are simply not available on CloudABI. I don't want to turn the official libuv codebase into an `#ifdef` fest.
- There is nothing available like kqueue, epoll, etc. I want libuv to be the core API against which event-driven software should be written; "libuv is my kqueue". At the same time I'd like to investigate whether I can implement this in a more async fashion, where operations can be enqueued within the kernel. We'll see.

Long story short, I'm trying to build Node.js against cloudlibc's of libuv. This is going pretty well so far, except that I run into various compilation errors for which the best course of action is debatable:

- Uses of `struct uv_*_s`, as opposed to `uv_*_t`.
- Uses of `UV_ERRNO_MAP()`.
- Relying on definitions that are available through `<uv.h>` implicitly, such as constants from `<sys/socket.h>`, `<netinet/in.h>`, `<netdb.h>`, etc.

My question is, are these things that Node.js should stop doing, or is it fair to say that that's simply part of what libuv offers? If it's the latter, do you want me to take a stab at documenting this?

**Comment by bnoordhuis** [comment]
> Uses of UV_ERRNO_MAP().

We export that intentionally (because I hate typing.)

> Uses of `struct uv_*_s`, as opposed to `uv_*_t`.

The `struct uv_loop_s` forward declaration in node.h is to avoid pulling in uv.h for a single type (which is only used once and then only in an API intended for embedders.)

I found only one other place where a `uv_*_s` type was used and I filed a PR to fix that.

> Relying on definitions that are available through <uv.h> implicitly, such as constants from <sys/socket.h>, <netinet/in.h>, <netdb.h>, etc.

Node.js should not be doing that, no. :-)

**Comment by EdSchouten** [comment]
Hallo Ben,

Thanks for replying so quickly! I'll make sure to add `UV_ERRNO_MAP()` and `struct uv_*_s` to my implementation and send out a pull request for documenting those. At the same time, I'll send pull requests to Node.js for fixing some of those implicit header dependencies.

**Comment by EdSchouten** [comment]
I also noticed that Node.js's `src/stream_wrap.h:70` uses `uv_pipe_t::ipc`. Ben, do you think we should also start making this part of the documented API?

**Comment by bnoordhuis** [comment]
It's not considered private.  That said, there is no reason node.js couldn't just remember the type of handle and that's arguably a little cleaner.

**Comment by cjihrig** [comment]
It seems like this can be closed. There may be some things that can be addressed in Node.js, but there's no need for an issue here to remain open. Please reopen if I'm wrong.
